testplan/testing/base.py

```diff
@@ -722,3 +761,83 @@ def aborting(self):
         if self._test_process is not None:
             kill_process(self._test_process)
             self._test_process_killed = True
+
+    def dry_run(self):
+        """
+        Return an empty report skeleton for this Test including all
+        testsuites, testcases etc. hierarchy. Does not run any tests.
+        """
+        result = super(ProcessRunnerTest, self).dry_run()
+        report = result.report
+
+        testsuite_report = TestGroupReport(
+            name="ProcessChecks",
+            category=ReportCategories.TESTSUITE,
+            uid="ProcessChecks",
+        )
+
+        testcase_report = TestCaseReport(
+            name="ExitCodeCheck", uid="ExitCodeCheck", suite_related=True,
+        )
+        testsuite_report.append(testcase_report)
+        report.append(testsuite_report)
+
+        return result
+
+    def run_testcases_iter(self, testsuite_pattern="*", testcase_pattern="*"):
+        """
+        For a Test to be run interactively, it must implement this method.
+
```

Comment:
I have misgivings about this. The docstring suggests this should be decorated `@abc.abstractmethod`, which means the class needs to use metaclass `abc.ABCMeta`, which may not be something you want to do in this preexisting class.

But since overriding this method is required for running `ProcessRunnerTest`-derived tests in interactive mode, how about making this more obvious by moving it to a new class: 

```py
class InteractiveProcessRunnerTest(ProcessRunnerTest, metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def run_testcases_iter(self, testsuite_pattern="*", testcase_pattern="*"):
        ...
```

That, or write this in such a way that the implementor of a class doesn't need to know to override this method (i.e. it will "just work" for whatever subclass derives from it).

Else I can imagine a future test-writer deriving from `ProcessRunnerTest` > trying to run it in interactive mode > it doesn't work > test-writer does not read the docstrings of all `ProcessRunnerTest` methods > test-writer stays frustrated and confused.

Reply:
To be clear, run_testcases_iter is an abstract method defined in the Test base class, this here is the concrete implementation. Maybe I should update the docstring to make that clearer.

I agree we should refactor Test and other abstract base classes to use the ABC meta-class, but that is not really in the scope of what this PR is trying to address.

I don't personally see the advantage of creating yet another base class specifically for interactive usage.

Reply:
I see, yes the docstring makes it seem like it must be overridden by any derived classes.

And agreed we way too many base classes as it is, but if something different must be done in order for the test to be run interactively I think it's worth it to make a new class, i.e. the benefits of making it obvious that there's additional work to be done by the implementor outweighs the burden of dealing with another base class.

Reply:
Since the goal I'm trying to get at is to make it obvious to the implementor that they need to do something extra in order to get their test to run interactively, a alternative to a new class could me an error message (either on the Python side or in the UI) that says something like "Cannot run interactively - implement the method `run_testcases_iter` in <insert name of class missing `run_testcases_iter`>. For an example implementation, see `testplan.testing.base.ProcessRunnerTest.run_testcases_iter`."

That is, if it absolutely must be overridden by a derived class.

Reply:
I agree in general that we should use the abc module to mark abstract classes and methods. However that applies to many many base classes, not just this ProcessTestRunner class. So though I think the change you're proposing is outside the scope of this PR, I think we should raise an issue to address it in future to all base classes that have abstract methods, not just this particular one.

Reply:
Here you go, I've raised an issue to address: https://github.com/Morgan-Stanley/testplan/issues/394